src/Razor/src/Microsoft.CodeAnalysis.Razor.Workspaces/Logging/ILoggerExtensions.cs

```diff
@@ -0,0 +1,151 @@
+    {
+        Log(logger, LogLevel.Warning, exception: null, message, args);
+    }
+
+    public static void LogWarning(this ILogger logger, Exception exception, string? message, params object?[] args)
+    {
+        Log(logger, LogLevel.Warning, exception, message, args);
+    }
+
+    public static void LogError(this ILogger logger, string? message, params object?[] args)
+    {
+        Log(logger, LogLevel.Error, exception: null, message, args);
+    }
+
+    public static void LogError(this ILogger logger, Exception exception, string? message, params object?[] args)
+    {
+        Log(logger, LogLevel.Error, exception, message, args);
+    }
+
+    public static void LogCritical(this ILogger logger, Exception exception, string? message, params object?[] args)
+    {
+        Log(logger, LogLevel.Critical, exception, message, args);
+    }
+
+    private static void Log(ILogger logger, LogLevel level, Exception? exception, string? message, object?[] args)
+    {
+        logger.Log(level, (message, args), exception, FormatMessage);
+    }
+
+    private static string FormatMessage((string? message, object?[] args) s, Exception? exception)
+    {
+        var (format, args) = s;
+        if (format is null)
+        {
+            return "[null]";
+        }
+
+        // from: https://github.com/dotnet/runtime/blob/ec4437be46d8b90bc9fa6740c556bd860d9fe5ab/src/libraries/Microsoft.Extensions.Logging.Abstractions/src/LogValuesFormatter.cs
+        using var _ = StringBuilderPool.GetPooledObject(out var vsb);
+        var scanIndex = 0;
+        var endIndex = format.Length;
+
+        var count = 0;
+
+        while (scanIndex < endIndex)
+        {
+            var openBraceIndex = FindBraceIndex(format, '{', scanIndex, endIndex);
+            if (scanIndex == 0 && openBraceIndex == endIndex)
+            {
+                // No holes found.
+                return format;
+            }
+
+            var closeBraceIndex = FindBraceIndex(format, '}', openBraceIndex, endIndex);
+
+            if (closeBraceIndex == endIndex)
+            {
+                vsb.Append(format, scanIndex, endIndex - scanIndex);
+                scanIndex = endIndex;
+            }
+            else
+            {
+                // Format item syntax : { index[,alignment][ :formatString] }.
+                var formatDelimiterIndex = format.AsSpan(openBraceIndex, closeBraceIndex - openBraceIndex).IndexOfAny(',', ':');
+                formatDelimiterIndex = formatDelimiterIndex < 0 ? closeBraceIndex : formatDelimiterIndex + openBraceIndex;
+
+                vsb.Append(format, scanIndex, openBraceIndex - scanIndex + 1);
+                vsb.Append(count++);
+                vsb.Append(format, formatDelimiterIndex, closeBraceIndex - formatDelimiterIndex + 1);
+
+                scanIndex = closeBraceIndex + 1;
+            }
+        }
+
+        return string.Format(vsb.ToString(), args);
```

Comment:
It feels strange to call `string.Format` here, which will just use another `StringBuilder` to call `AppendFormat`.